frontend/controller/service-worker.js

```diff
@@ -62,17 +62,31 @@ sbp('sbp/selectors/register', {
 
     const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
     const existingSubscription = await registration.pushManager.getSubscription()
+    const messageToPushServerIfSocketOpen = (msgPayload) => {
+      // make sure the websocket client is not in the state of CLOSING, CLOSED before sending a message.
+      // (context: https://github.com/okTurtles/group-income/pull/1770#discussion_r1439005731)
+
+      const readyState = pubsub.socket.readyState
+      const sendMsg = () => pubsub.socket.send(createMessage(
+        REQUEST_TYPE.PUSH_ACTION,
+        msgPayload
+      ))
+
+      if (readyState === WebSocket.CLOSED || readyState === WebSocket.CLOSING) {
+        pubsub.socket.addEventListener('open', function openHandler () {
+          sendMsg()
+          pubsub.socket.removeEventLister('open', openHandler)
```

Comment:
This `open` handler is never going to be called, since a web socket cannot be reused once closed. That's why the client uses a new socket for every connection, and automatically removes and re-adds its socket event handlers.

Such custom handlers can be registered under the 'handlers' option in `createClient`, where the client is created (you already know the `messageHandlers` option which is very similar).

Reply:
@snowteamer
>  a web socket cannot be reused once closed

 That's a good point that I missed while working on the previous commit. Will make another update that resolves this.

Reply:
@snowteamer I realised that there is already a sbp event `PUBSUB_RECONNECTION_SUCCEEDED` that you created, which is emitted when the WebSocket has lost the network connection and then is reconnected. This seems to be useful in this case so decided to resolve the mentioned issue using it. ([here](https://github.com/okTurtles/group-income/pull/1799/commits/0327c551132cb4db8deeb72754d5cf538db37a9b))